`glue::task::spawn` shouldn't panic when the entire wasm program is running on a WebWorker

Recently I'm writing a wasm program using `tokio_with_wasm` that runs in a standalone WebWorker (meaning that I called init functions from `wasm-bindgen` in the worker). This program runs entirely in the worker, it only communicates with the main thread through `onmessage`/`postMessage`. 

The program will just panic when calling `spawn`, at https://github.com/cunarist/tokio-with-wasm/blob/b4171d75a899c02cea4afc3a3f523c3809959cae/package/src/glue/task/mod.rs#L181

From https://github.com/cunarist/tokio-with-wasm/pull/27#issuecomment-2933771700 I can see that `tokio_with_wasm` doesn't allow spawning in a blocking task to avoid undefined reference to JS objects. But in my situation, wasm program running in the WebWorker is not in a blocking task, and all JS objects it uses just stay in the Worker. `spawn` shouldn't panic under this situation.

The possible fix may be setting a flag to indicate whether the program is in a blocking_task, and decides to panic only when the progra was in a blocking_task.

--- a/package/Cargo.toml
+++ b/package/Cargo.toml
@@ -36,5 +36,4 @@ web-sys = { version = "0.3.70", features = [
     'Blob',
     "BlobPropertyBag",
     'Url',
-    'Window',
 ] }

--- a/package/src/glue/common/thread_check.rs
+++ b/package/src/glue/common/thread_check.rs
@@ -1,9 +1,19 @@
+use js_sys::global;
 use std::cell::LazyCell;
-use web_sys::window;
+use wasm_bindgen::JsValue;
+
+/// The name of a JS object
+/// that is only present in the blocking thread.
+pub static BLOCKING_KEY: &str = "isBlockingTokioThread";
 
 thread_local! {
-  pub static IS_MAIN_THREAD: LazyCell<bool> =
-    LazyCell::new(|| window().is_some());
+  pub static IS_MAIN_THREAD: LazyCell<bool> = LazyCell::new(|| {
+    let global_obj = global();
+    let is_blocking_thread =
+      js_sys::Reflect::has(&global_obj, &JsValue::from_str(BLOCKING_KEY))
+        .unwrap_or(false);
+    !is_blocking_thread
+  });
 }
 
 pub fn is_main_thread() -> bool {

--- a/package/src/glue/task/pool.rs
+++ b/package/src/glue/task/pool.rs
@@ -1,4 +1,4 @@
-use crate::{LogError, now};
+use crate::{BLOCKING_KEY, LogError, now};
 use js_sys::{Array, JsString, Object, Reflect, eval, global};
 use std::cell::RefCell;
 use std::collections::VecDeque;
@@ -77,6 +77,7 @@ impl WorkerPool {
       "
       import init, * as wasmBindings from '{}';
       globalThis.wasmBindings = wasmBindings;
+      globalThis.{BLOCKING_KEY} = true;
       self.onmessage = event => {{
         let initialised = init(event.data).catch(err => {{
           // Propagate to main `onerror`:

